### PR TITLE
MS-ILLM models and training scripts

### PR DESCRIPTION
**Pull request by mmuckley**
This pull request implements, the generative image compression model from the following paper:

MJ Muckley, A El-Nouby, K Ullrich, H Jegou, J Verbeek.
[Improving Statistical Fidelity for Neural Image Compression with Implicit Local Likelihood Models](https://openreview.net/forum?id=iUspLfxpWC).
In *ICML*, 2023.

The PR includes:
- A recursive bottleneck structure for implementing VQ-VAEs.
- A new VQ-VAE model, based on the VQ-GAN paper, that has XCiT attention for improved memory efficiency.
- A modification of the OASIS U-Net discriminator that clips the discriminator in the upsampling path corresponding to the VQ-VAE latent space.
- A PyTorch Lightning module for HiFiC-style training of hyperprior models for targeted bitrates.
- A PyTorch Lightning module for fine-tuning the HiFiC-autoencoders with MS-ILLM

Installation instructions and basic usage are covered in the README in the `illm` project folder

### CODE REVIEW

File: neuralcompression/layers/_project_latent.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        output_dim: The output dimension for the projection.
+        child: A child operation to apply between forward and inverse
+            projections.
+    """
+
+    def __init__(
+        self, input_dim: int, output_dim: int, child: Optional[nn.Module] = None
+    ):
+        super().__init__()
+        self.child = child
```
Comment (QuentinDuval) [inline neuralcompression/layers/_project_latent.py:33]: Same comment regarding the linked list in disguise, in that case it is not a linked list, but another data structure in disguise - but looking at the forward function, an nn.Sequential built by a factory function could do the job ?

File: neuralcompression/layers/_xca.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+
+
+class XCA(nn.Module):
```
Comment (QuentinDuval) [inline neuralcompression/layers/_xca.py:28]: I love this stuff, so sad the paper was not used more

File: neuralcompression/models/_unet_discriminator.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+def apply_conv_norm(module: nn.Module, norm_type: str = "identity") -> nn.Module:
+    _verify_norm_type(norm_type)
+    if norm_type == "spectral":
+        return spectral_norm(module)
+    else:
+        return module
+
+
+def build_norm_layer(chans: int, norm_type: str) -> nn.Module:
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:32]: In Omnivision, we avoid this kind of functions by requiring targets in configuration files (but it is waaay more verbose)

File: neuralcompression/models/_unet_discriminator.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        self.left_shortcut = nn.Sequential(
+            left_shortcut_conv,
+            nn.AvgPool2d(2),
+        )
+
+        self.right_block: nn.Module
+        self.right_shortcut: Union[nn.Module, Callable]
+        if clip_upsample is True:
+            self.right_block = nn.Identity()
+            self.right_shortcut = lambda x: 0.0
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:121]: should it return instead a torch.tensor ? I guess it will be automatically casted to it anyway ?

File: neuralcompression/models/_unet_discriminator.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        bias=bias,
+    )
+
+
+class _UnetLevel(nn.Module):
+    def __init__(
+        self,
+        in_planes: int,
+        down_planes: int,
+        child: nn.Module,
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:65]: Same comment as before: the `child` is an implicit data structure in disguise that could be replaced with List, nn.Sequential or nn.ModuleList

File: neuralcompression/models/_unet_discriminator.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            )
+
+    def forward(self, image: Tensor) -> Tensor:
+        image = self.child(self.left_block(image) + self.left_shortcut(image))
+        return self.right_block(image) + self.right_shortcut(image)
+
+
+class UnetDiscriminator(Discriminator):
+    """
+    OASIS U-Net Discriminator
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:158]: Niiiice :)

File: neuralcompression/models/_unet_discriminator.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+    You Only Need Adversarial Supervision for Semantic Image Synthesis
+    V Sushko, E Schönfeld, D Zhang, J Gall, B Schiele, A Khoreva
+
+    It includes upsampling path modifications to calculate the discriminator
+    used in neural compression paper:
+
+    Improving Statistical Fidelity for Neural Image Compression with Implicit
+    Local Likelihood Models
+    MJ Muckley, A El-Nouby, K Ullrich, H Jégou, J Verbeek
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:171]: "MJ Muckley" is very classy :)

File: neuralcompression/models/_unet_discriminator.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+        clip_upsample = False
+        for count, chan in enumerate(reversed(self.channels[1:-1])):
+            if count == num_upsamples:
+                clip_upsample = True
+
+            model = _UnetLevel(
+                in_planes=chan,
+                out_planes=(64 if count == num_upsamples - 1 else None),
+                down_planes=lastchan,
+                child=model,
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:224]: Ok, I understand now how the `child` is being used - you build a recursive data structure by wrapping a UNet with another Unet - it's smart - I still wonder if it can be done better using explicit rather than implicit data structures

File: neuralcompression/models/_unet_discriminator.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                context_out,
+                kernel_size=3,
+                padding=1,
+                padding_mode="reflect",
+            ),
+            nn.LeakyReLU(negative_slope=0.2, inplace=True),
+        )
+
+    @property
+    def is_conditional(self) -> bool:
```
Comment (QuentinDuval) [inline neuralcompression/models/_unet_discriminator.py:322]: I like those kind of flags too, instead of isinstance

File: neuralcompression/optim/_ramp_const_cosine.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+
+from torch.optim import Optimizer
+from torch.optim.lr_scheduler import LambdaLR
+
+
+class RampConstCosineLRSchedule(LambdaLR):
```
Comment (QuentinDuval) [inline neuralcompression/optim/_ramp_const_cosine.py:12]: Did not know about this one : is it proposed in a paper ? (adding citation ?) or is it something that worked well in practice for you ?

File: projects/illm/README.md [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    +pretrained_autoencoder=example \
+    distortion_loss=mse_lpips \
+    distortion_loss.mse_param=150.0 \
+    distortion_loss.lpips_param=1.0 \
+    distortion_loss.backbone=alex \
+    hydra.launcher.timeout_min=4320 \
+    +discriminator=condunet_ch1025_factor8_context220 \
+    +lightning_module.generator_weight=0.0005 \
+    lightning_module.noisy_context_steps=0 \
+    lightning_module.distortion_lam=1.0,0.5,0.25,0.125,0.0625,0.03125,0.015625,0.0078125 \
```
Comment (QuentinDuval) [inline projects/illm/README.md:116]: Is this running a sweep in behind ? Or is it specifying a list? In any case, I think you could make this easier by encapsulating things like that into configs that have a name indicating what you are trying to do.

For instance instead of all the lines about `distortion.*=*`, you could have `+distortion=XXX`.

File: projects/illm/default_transforms.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    ToTensor,
+)
+
+
+def default_train_transform(image_size: int) -> Compose:
+    return Compose(
+        [
+            RandomChoice(
+                [
+                    RandomCrop(
```
Comment (QuentinDuval) [inline projects/illm/default_transforms.py:23]: I tend to avoid nesting to much and having that amount of indent for transforms, mostly by assigning to intermediate variables - not sure what's best thought

File: projects/illm/gan_compression_module.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+            return self.discriminator(images, context)
+        else:
+            return self.discriminator(images)
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        full_images: Tensor
+        output: HyperpriorOutput
```
Comment (QuentinDuval) [inline projects/illm/gan_compression_module.py:82]: Oh interesting ! You declare the types at the start ?

File: projects/illm/gan_compression_module.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+        scheds = self.lr_schedulers()
+        if len(opts) == 3:
+            model_opt, disc_opt, quantile_opt = opts
+            model_sched, disc_sched, quantile_sched = scheds
+        else:
+            model_opt, disc_opt = opts
+            model_sched, disc_sched = scheds
+            quantile_opt = None
+
+        # if doing Monte Carlo sampling, split the batch for the discriminator
```
Comment (QuentinDuval) [inline projects/illm/gan_compression_module.py:93]: Each part of the process could be split into methods - one algorithm I use is that each time I write a comment, I try to write a function with the same name instead

File: projects/illm/gan_compression_module.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+        # get the discriminator/generator targets
+        if self.latent_projector is not None:
+            with torch.no_grad():
+                fake_target = self.latent_projector(model_images)
+                real_target = self.latent_projector(disc_images)
+        else:
+            real_target = torch.tensor(1, dtype=torch.long)
+            fake_target = torch.tensor(1, dtype=torch.long)
+
+        ########### GENERATOR STEP ###########
```
Comment (QuentinDuval) [inline projects/illm/gan_compression_module.py:112]: same as above: could extract a function named - generator step

File: projects/illm/gan_compression_module.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+        if self.gradient_clip_val is not None:
+            self.clip_gradients(
+                model_opt,
+                gradient_clip_val=self.gradient_clip_val,
+                gradient_clip_algorithm=self.gradient_clip_algorithm,
+            )
+        model_opt.step()
+        model_sched.step()
+
+        ########### DISCRIMINATOR STEP ###########
```
Comment (QuentinDuval) [inline projects/illm/gan_compression_module.py:154]: Same as above - could extract a function named `_discriminator_step`

File: projects/illm/train.py [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+from torch.utils.data import DataLoader
+from torchvision.utils import make_grid
+
+from neuralcompression.loss_fn import TargetRateConfig
+from neuralcompression.models import HyperpriorAutoencoderBase
+
+LOGGER = logging.getLogger(__file__)
+
+
+class WandbImageCallback(pl.Callback):
```
Comment (QuentinDuval) [inline projects/illm/train.py:38]: Note: The wandb part is to me a major block to using the code if it cannot be removed with just a single flag.

**Review by QuentinDuval — APPROVED** [review]

File: .gitignore [hunk 18]
```diff
@@ -157,3 +157,6 @@ docs/source/examples
 
 # setuptools_scm
 neuralcompression/version.py
+
+# logs created by hydra
```
[the remark this answers is not in the document]
Comment (mmuckley) [reply .gitignore:161]: Generally agree, but in this case for new research users who are less familiar with this paradigm I want the logs to appear right where the entry point is. After that they can move it if they want. Of course, our internal code has the logs in another directory.

File: neuralcompression/layers/_normalize_latent.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    dimension prior to a subsequent child operation.
+
+    Args:
+        child: A child operation to follow normalization within an overall
+            recursive tree.
+    """
+
+    def __init__(self, child: Optional[nn.Module] = None):
+        super().__init__()
+        self.child = child
```
[the remark this answers is not in the document]
Comment (mmuckley) [reply neuralcompression/layers/_normalize_latent.py:28]: I'll add this as a TODO. At the moment I don't think compute would change much at all (the operations are too heavy), but the refactor may require some consideration.

File: projects/illm/gan_compression_module.py (hunk 13 above)
Comment (mmuckley) [reply projects/illm/gan_compression_module.py:82]: Ah yes, this is just a way to help my poor little VS Code give me the right functions and types for each variable.

File: neuralcompression/optim/_ramp_const_cosine.py (hunk 10 above)
Comment (mmuckley) [reply neuralcompression/optim/_ramp_const_cosine.py:12]: Just something that worked well in practice. We document it in the appendix. But I think it's sort of a common knowledge thing for NLP at this point.

File: neuralcompression/models/_unet_discriminator.py (hunk 5 above)
Comment (mmuckley) [reply neuralcompression/models/_unet_discriminator.py:65]: Just a personal preference, I've usually liked this sort of tree structure because it discards intermediate variables as you progress back up the tree. Usually with the list implementation you keep everything around until the end. Saves a small amount of memory.

File: neuralcompression/models/_unet_discriminator.py (hunk 3 above)
Comment (mmuckley) [reply neuralcompression/models/_unet_discriminator.py:32]: Oh I do think that would be better, but might put too much hydra into here.

File: projects/illm/README.md (hunk 11 above)
Comment (mmuckley) [reply projects/illm/README.md:116]: In this case it's meant to be a didactic example for how to do hydra sweeps. Omnivision has a sort of experiment-based config structure, but here we want to leverage the launcher for job arrays to sweep over a whole bunch of changes.

File: projects/illm/default_transforms.py (hunk 12 above)
Comment (mmuckley) [reply projects/illm/default_transforms.py:23]: I removed a little bit of indentation.

File: projects/illm/gan_compression_module.py (hunk 15 above)
Comment (mmuckley) [reply projects/illm/gan_compression_module.py:119]: I worked on this a little bit, but opted against it for a couple of reasons:

1. There are a bunch of inter-related variables, so we don't save very much by functionalizing.
2. Philosophically I think the training step should give a very good picture of the global state for calculating the loss, and to give this picture distribution should be avoided.

File: projects/illm/train.py (hunk 17 above)
Comment (mmuckley) [reply projects/illm/train.py:45]: I added a `try`/`except` for now so the code should potentially run if the user doesn't have it, but this should probably be configurable.
